### ISSUE TITLE
Android/IOS Version

### ISSUE THREAD

**Issue by memtech3** [flxzt/rnote#390]
It would be super cool if there was an Android version of the full rnote. In another feature request (#245) , a commenter mentioned compiling to webassembly to bring android support. I think as long as the compiled webassembly version was distributed as a PWA that could be installed and used offline that would satisfy this feature request

Some context: I am a college student and my day to day loadout is a laptop (non touch) and android tablet with stylus (Galaxy Tab S6 Lite) as it is more convenient than carrying my wired Wacom tablet around to classes.

**Comment by pktiuk** [comment]
As a first step of checking doability of this task it should be checked whether base libraries (like GTK4) can be compiled into webassembly.

To check convergence and web browser compatibility you can run the current application with [GTK4 Broadway](https://docs.gtk.org/gtk4/broadway.html) backend.  

> The GDK Broadway backend provides support for displaying GTK applications in a web browser, using HTML5 and web sockets.

[phoronix news about broadway](https://www.phoronix.com/news/GTK4-Broadway-Being-Used)

**Comment by Shahin-rmz** [comment]
I can help with [Kivy](https://kivy.org/) if this is helpful!

**Comment by pktiuk** [comment]
@Shahin-rmz How using Kivy is supposed to help? Kivy is Python framework, but Rnote is written with Rust and GTK4.    
Do I miss something?

**Comment by dbrandonjohnson** [comment]
I'm uncertain if this would help: https://github.com/tauri-apps/tauri-mobile

**Comment by Beryesa** [comment]
I wonder if there can be a file-format compatible side project for android, I haven't taken a look at the source yet but if core logic and gtk bindings are modularly implemented, maybe a rust+flutter branch can be maintained without too much effort. (High likely much easier than trying to get any gtk on android?)

I don't have enough knowledge in either of them but in the case of no interest/time, perhaps I can try to step up and save all the students from searching endlessly for a proper but cross platform foss annotation software, myself included, at some time in the future (next year?) :3

**Comment by mistune** [comment]
Anyone able to build this in a termux proot environment? I can't seem to get this done in Debian due to some dependencies being outdated.

**Comment by DeltriDev** [comment]
How is this going? Is there a android/web version in development?

**Comment by mistune** [comment]
Trying to build this in version 0.8.2 where most of the dependencies in termux are available without manually building the dependencies. I'm currently stuck with the package gettext-sys. I pasted the logs here, hopefully someone more knowledgeable can help debug this. https://github.com/termux/termux-packages/issues/17655#issuecomment-2016687563

**Comment by DeltriDev** [comment]
> Trying to build this in version 0.8.2 where most of the dependencies in termux are available without manually building the dependencies. I'm currently stuck with the package gettext-sys. I pasted the logs here, hopefully someone more knowledgeable can help debug this. [termux/termux-packages#17655 (comment)](https://github.com/termux/termux-packages/issues/17655#issuecomment-2016687563)
> 
> EDIT: Tried compiling in a Fedora environment but seems Termux crashes during the last package to be compiled.
> 
> EDIT2: Somebody compiled this for aarch64 and just installed it from there.
> 
> EDIT3: Running this on Termux-x11 + Hardware Acceleration through virgl and I am getting close to native performance on my tablet.

that's great! is there a build somewhere or do i have to build this myself? i would rather not

**Comment by mistune** [comment]
@DeltriDev I used the build here https://fedora.pkgs.org/39/terra/rnote-0:0.10.1-1.fc39.aarch64.rpm.html.

**Comment by PhilDevProg** [comment]
If someone makes a Flutter version in the future, maybe the [flutter_rust_bridge](https://github.com/fzyzcjy/flutter_rust_bridge) could be helpful for reusing code instead of having to rewrite everything.

**Comment by PhilDevProg** [comment]
Someone's working on an Android backend for Gtk: https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/7555

**Comment by sound-friend** [comment]
man i would love to see something like this just so i can check my notes on my phone!!

**Comment by EDLLT** [comment]
https://github.com/termux/termux-packages/pull/21524

I managed to get RNote to work in Android natively(no proot/chroot) through termux with hardware acceleration

**Comment by sound-friend** [comment]
could you let us know how you did it? what's the performance like?

**Comment by EDLLT** [comment]
> could you let us know how you did it? what's the performance like?

Performance is much more better than  chroot and proot, especially for large rnote files

Also it responds correctly to touch for panning and zooming, zoom and pen for writing

I did it by patching rnote and several crates to compile against Android's bionic libc(Check PR above)

Soon enough once the merging with termux packages is done, installing rnote will be as easy as "pkg install rnote"

If you're in the termux discord or rnote dev matrix channel then you'll find that I've sent a short video there demonstrating rnote

**Comment by mistune** [comment]
> [termux/termux-packages#21524](https://github.com/termux/termux-packages/pull/21524)
> 
> I managed to get RNote to work in Android natively(no proot/chroot) through termux with hardware acceleration

Is this in the repos now? Doing a `pkg search rnote` does not show rnote yet

**Comment by relma2** [comment]
The pull request to get **rnote** onto the termux repos has been merged! I am running `pkg install rnote` in termux right now

**Comment by relma2** [comment]
Ok, running into several issues. Will go home and resume troubleshooting.

**Comment by sound-friend** [comment]
when i run it it says it can't find the package... interesting

**Comment by relma2** [comment]
> when i run it it says it can't find the package... interesting

Install the x11 repo first with `pkg install x11-repo`

**Comment by relma2** [comment]
@sound-friend , @EDLLT , and I were in the termux discord the other night getting EDLLT's packaging of Rnote to work in Termux. It... kind of works, the application loads, but there are several issues. The steps I did to get it """working""" were:

1. Install [Termux](https://f-droid.org/en/packages/com.termux/) from F-Droid (NOT the Google Play store!!)
2. Install the [Termux-X11 Nightly](https://github.com/termux/termux-x11?tab=readme-ov-file#setup-instructions) Apk fron Github
3. In Termux, run `pkg install x11-repo` to enable the X11 Repo of packages
4. Run `pkg install rnote xfce4` to install Rnote and the XFCE Desktop Environment
5. Run the XFCE Desktop Environment with `termux-x11 :1 -xstartup "dbus-launch --exit-with-session xfce4-session"`
6. Back in Termux, install Hardware acceleration with `pkg install mesa-vulkan-icd-freedreno-dri3`
7. Open the Termux X11 App. You should see an XFCE Desktop environment running with a little mouse wallpaper. 
8. In a Terminal WITHIN THE XFCE ENVIRONMENT (not termux), run `MESA_LOADER_DRIVER_OVERRIDE=zink rnote`
9. Rnote opens! You can draw on it, but maximizing the window causes the app to segfault. Pen pressure also does not work.

Tl;dr -- this approach is nowhere near a functioning "android version" of Rnote. We should probably look into compiling GTK natively  on Android or something; this approach uses too many workarounds and the app doesen't have full functionality and segfaults.

**Comment by PhilDevProg** [comment]
> @sound-friend , @EDLLT , and I were in the termux discord the other night getting EDLLT's packaging of Rnote to work in Termux. It... kind of works, the application loads, but there are several issues. The steps I did to get it """working""" were:
> 
> 1. Install [Termux](https://f-droid.org/en/packages/com.termux/) from F-Droid (NOT the Google Play store!!)
> 2. Install the [Termux-X11 Nightly](https://github.com/termux/termux-x11?tab=readme-ov-file#setup-instructions) Apk fron Github
> 3. In Termux, run `pkg install x11-repo` to enable the X11 Repo of packages
> 4. Run `pkg install rnote xfce4` to install Rnote and the XFCE Desktop Environment
> 5. Run the XFCE Desktop Environment with `termux-x11 :1 -xstartup "dbus-launch --exit-with-session xfce4-session"`
> 6. Back in Termux, install Hardware acceleration with `pkg install mesa-vulkan-icd-freedreno-dri3`
> 7. Open the Termux X11 App. You should see an XFCE Desktop environment running with a little mouse wallpaper. 
> 8. In a Terminal WITHIN THE XFCE ENVIRONMENT (not termux), run `MESA_LOADER_DRIVER_OVERRIDE=zink rnote`
> 9. Rnote opens! You can draw on it, but maximizing the window causes the app to segfault. Pen pressure also does not work.
> 
> Tl;dr -- this approach is nowhere near a functioning "android version" of Rnote. We should probably look into compiling GTK natively  on Android or something; this approach uses too many workarounds and the app doesen't have full functionality and segfaults.

There seems to be a problem with the Zink driver on my device (Pixel 8a):
https://github.com/user-attachments/assets/a22b6b0f-e744-471d-85e9-5129678f909c

When I run Rnote without `MESA_LOADER_DRIVER_OVERRIDE=zink, this doesn't happen.

**Comment by relma2** [comment]
> When I run Rnote without `MESA_LOADER_DRIVER_OVERRIDE=zink, this doesn't happen.

Yeah, zink has a lot of issues. When I run without zink, I don't get the segfault on maximizing the window

**Comment by EDLLT** [comment]
> > @sound-friend , @EDLLT , and I were in the termux discord the other night getting EDLLT's packaging of Rnote to work in Termux. It... kind of works, the application loads, but there are several issues. The steps I did to get it """working""" were:
> > 
> > 1. Install [Termux](https://f-droid.org/en/packages/com.termux/) from F-Droid (NOT the Google Play store!!)
> > 2. Install the [Termux-X11 Nightly](https://github.com/termux/termux-x11?tab=readme-ov-file#setup-instructions) Apk fron Github
> > 3. In Termux, run `pkg install x11-repo` to enable the X11 Repo of packages
> > 4. Run `pkg install rnote xfce4` to install Rnote and the XFCE Desktop Environment
> > 5. Run the XFCE Desktop Environment with `termux-x11 :1 -xstartup "dbus-launch --exit-with-session xfce4-session"`
> > 6. Back in Termux, install Hardware acceleration with `pkg install mesa-vulkan-icd-freedreno-dri3`
> > 7. Open the Termux X11 App. You should see an XFCE Desktop environment running with a little mouse wallpaper. 
> > 8. In a Terminal WITHIN THE XFCE ENVIRONMENT (not termux), run `MESA_LOADER_DRIVER_OVERRIDE=zink rnote`
> > 9. Rnote opens! You can draw on it, but maximizing the window causes the app to segfault. Pen pressure also does not work.
> > 
> > Tl;dr -- this approach is nowhere near a functioning "android version" of Rnote. We should probably look into compiling GTK natively  on Android or something; this approach uses too many workarounds and the app doesen't have full functionality and segfaults.
> 
> There seems to be a problem with the Zink driver on my device (Pixel 8a):
> https://github.com/user-attachments/assets/a22b6b0f-e744-471d-85e9-5129678f909c
> 
> When I run Rnote without `MESA_LOADER_DRIVER_OVERRIDE=zink, this doesn't happen.

Zink hardware acceleration is for Adreno GPUs, from my Google search, the Pixel 8a has a Mali GPU

In your case, running without zink makes it use llvmpipe which is the CPU(cpus are not designed for rendering)

Instead, you could setup VirGL(Virtual opengl) for termux, it's a bit more complicated than zink so you'd need to ask people in the termux discord server)

**Comment by PhilDevProg** [comment]
> Zink hardware acceleration is for Adreno GPUs, from my Google search, the Pixel 8a has a Mali GPU
> 
> In your case, running without zink makes it use llvmpipe which is a software rasterizer to run using the CPU(cpus are not designed for rendering so it's slow)
> 
> Instead, you could setup VirGL(Virtual opengl) for termux, it's a bit more complicated than zink so you'd need to ask people in the termux discord server)

Thanks for that tip. I got it working by using the following commands:
`pkg install virglrenderer-android`
`virgl_test_server_android &` and then started Rnote with this: `GALLIUM_DRIVER=virpipe MESA_GL_VERSION_OVERRIDE=4.0 rnote`

**Comment by mistune** [comment]
Tried the following on two devices. You might have to test which one will work for your device

### Installation

Zink

```
pkg update && pkg upgrade -y && pkg install rnote xfce4 glmark2 angle-android mesa-zink-dev virglrenderer-android virglrenderer-mesa-zink vulkan-loader-android xorg-xrandr 
```

Turnip

```
pkg update && pkg upgrade -y && pkg install rnote xfce4 glmark2 mesa-vulkan-icd-freedreno-dri3
```

### Start XFCE4

```
termux-x11 :1 -xstartup "dbus-launch --exit-with-session xfce4-session"
```

### Run program

Zink

```
GALLIUM_DRIVER=zink MESA_GL_VERSION_OVERRIDE=4.0 rnote
```

Turnip

```
MESA_LOADER_DRIVER_OVERRIDE=zink rnote
```

**Comment by GurrenLagann114** [comment]
Any update on this???

**Comment by PhilDevProg** [comment]
The Android backend for Gtk has just been merged making a port of Rnote much more feasible: https://gitlab.gnome.org/GNOME/gtk/-/merge_requests/7555

**Comment by Doublonmousse** [comment]
Taking a quick look at it, it's not (yet) usable for rnote yet.
- There seems to be pen support (so that's good, I've only looked at the code though, not tested with physical hardware)
- The only supported renderer is cairo which doesn't work with rnote (if you try to force this renderer with rnote on pc, it will give the pink rectangle bug, because of an unsupported feature in the cairo backend https://gitlab.gnome.org/GNOME/gtk/-/blob/main/gsk/gskrendernodeimpl.c#L6845)
- There are other usability issues for now (text fields don't work)

We need to wait on the gl/vulkan renderer support on android first to reconsider. Other blockers may come from the rust bindings packages (`cairo-rs`, `poppler-rs` and, more likely, `gettext-rs`)

**Comment by psomani16k** [comment]
Android 16 is getting a debian based terminal (basically a debian vm), could that be helpful?

**Comment by archerallstars** [comment]
@psomani16k 

Until it's finalized, there are many things remain to be seen. For example, is it able to run Flatpak?, how is the hardware acceleration?, how the pressure sensitivity is handled inside the VM, or the container for that matter?, if not, what about USB pass through for graphic drawing tablets?, etc.

I wouldn't have high hope at this point.

**Comment by psomani16k** [comment]
qt has working android support as far as I understand, how about porting rnote to qt? Is it viable? maybe getting it to work something like Krita?

**Comment by Intranox** [comment]
For those interested, I've created a rudimentary Android application that can at least view the Rnote format. I'll start by saying that I don't have the time or expertise to maintain or improve the app, but it works well for my personal use with my own notes (the vertical layout continues, even with a grid). If anyone in my situation would like to at least view these types of files on an Android device, this could be helpful. For those familiar with the language and Android SDK, this could be a good starting point for something even more stable. Feel free to fork it and adapt it to your needs or generalize it further. [https://github.com/Intranox/rnoteviewer-android](https://github.com/Intranox/rnoteviewer-android)

**Comment by SomeonetookSomeonetookmeep** [comment]
Very cool, I've finally gotten access to a phone with stylus, and love seeing this step to potentially doing Rnote on my phone natively!
(I cooould probably do the termux, but not right CPU for good speed iirc) Plus convenience of notes.

**Comment by KJ-Ly** [comment]
> For those interested, I've created a rudimentary Android application that can at least view the Rnote format. I'll start by saying that I don't have the time or expertise to maintain or improve the app, but it works well for my personal use with my own notes (the vertical layout continues, even with a grid). If anyone in my situation would like to at least view these types of files on an Android device, this could be helpful. For those familiar with the language and Android SDK, this could be a good starting point for something even more stable. Feel free to fork it and adapt it to your needs or generalize it further. https://github.com/Intranox/rnoteviewer-android

Great work! I've been putting together my own app (using some of your code) and it can now make some basic edits to .rnote files. I'll drop it here once it's more ready to share

**Comment by KJ-Ly** [comment]
Here is my Rnote editor for Android! It is able to view, edit and create Rnote files! Please note that it is not a true port as it doesn't use Rnote's engine but a rudimentary, reverse-engineered Kotlin implementation. (thank you again @Intranox for laying the groundwork for this)
https://github.com/KJ-Ly/Basic-Rnote-editor-Android
